File: sdk/typescript/libraries/bot-solutions/src/middleware/setSpeakMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Middleware, SendActivitiesHandler, TurnContext } from 'botbuilder';
import { Activity, ActivityTypes, ResourceResponse, Channels } from 'botframework-schema';
import { Element, js2xml, xml2js } from 'xml-js';

/**
 * Set Speech Synthesis Markup Language (SSML) on an Activity's Speak property with locale and voice input.
 */
export class SetSpeakMiddleware implements Middleware {
    private static readonly defaultLocale: string = 'en-us';
    private static readonly defaultVoiceFonts: Map<string, string> = new Map([
        ['de-de', 'Microsoft Server Speech Text to Speech Voice (de-DE, Hedda)'],
        ['en-us', 'Microsoft Server Speech Text to Speech Voice (en-US, Jessa24kRUS)'],
        ['es-es', 'Microsoft Server Speech Text to Speech Voice (es-ES, Laura, Apollo)'],
        ['fr-fr', 'Microsoft Server Speech Text to Speech Voice (fr-FR, Julie, Apollo)'],
        ['it-it', 'Microsoft Server Speech Text to Speech Voice (it-IT, Cosimo, Apollo)'],
        ['zh-cn', 'Microsoft Server Speech Text to Speech Voice (zh-CN, HuihuiRUS)']
    ]);
    private static readonly defaultChannels: Set<string> = new Set([Channels.DirectlineSpeech, Channels.Emulator]);
    private locale: string;
    private voiceFonts: Map<string, string>;
    private channels: Set<string>;
    private static readonly namespaceURI = 'https://www.w3.org/2001/10/synthesis';

    /**
     * Initializes a new instance of the SetSpeakMiddleware class.
     * @param locale If null, use en-US.
     * @param voiceFonts Map voice font for locale like en-US to "Microsoft Server Speech Text to Speech Voice (en-us, Jessa24kRUS).
     * @param channels Set SSML for these channels. If null, use DirectlineSpeech and Emulator.
     */
    public constructor(locale: string = '', voiceFonts: Map<string, string> = new Map<string, string>(), channels: Set<string> = new Set()) {
        this.locale = locale || SetSpeakMiddleware.defaultLocale;
        this.voiceFonts = voiceFonts.size > 0 ? voiceFonts : SetSpeakMiddleware.defaultVoiceFonts;
        this.channels = channels.size > 0 ? channels : SetSpeakMiddleware.defaultChannels;
    }

    /**
     * If outgoing Activities are messages and using the Direct Line Speech channel,
     * decorate the Speak property with an SSML formatted string.
     * @param context The Bot Context object.
     * @param next The next middleware component to run.
     */
    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        const handler: SendActivitiesHandler = async (
            ctx: TurnContext, activities: Partial<Activity>[], nextSend: () => Promise<ResourceResponse[]>
        ): Promise<ResourceResponse[]> => {
            activities.forEach((activity: Partial<Activity>): void => {
                switch (activity.type) {
                    case ActivityTypes.Message: {
                        activity.speak = this.getActivitySpeakText(activity);

                        if (activity.channelId !== undefined) {
                            if (this.channels.has(activity.channelId)) {
                                activity.speak = this.decorateSSML(activity);
                            }
                        }
                        break;
                    }
                    default:
                }
            });

            return await nextSend();
        };

        context.onSendActivities(handler);

        return next();
    }

    /**
     * Gets the speak text for the activity.
     * @param activity Outgoing bot Activity.
     * @returns speech text string value.
     */
    private getActivitySpeakText(activity: Partial<Activity>): string {
        // return speak or text value if they already exist in the activity
        const result: string | undefined = activity.speak || activity.text;
        if (result !== undefined) {
            return result;
        }

        // return speak value of first attachment if an attachment exists and has a speak value
        if (activity.attachments !== undefined && activity.attachments.length > 0) {
            const attachmentContent = activity.attachments[0].content;
            if (attachmentContent !== undefined && attachmentContent instanceof Object) {
                const contentObj = JSON.parse(attachmentContent);
                return contentObj['speak'].toString();
            }
        }

        return '';
    }

    /**
     * Formats an existing string to be formatted for Speech Synthesis Markup Language with a voice font.
     * @param activity Outgoing bot Activity.
     * @returns SSML-formatted string to be used with synthetic speech.
     */
    private decorateSSML(activity: Partial<Activity>): string {
        if (activity.speak === undefined || activity.speak.trim().length === 0) {
            return '';
        }

        let rootElement: Element | undefined = undefined;
        try {
            rootElement = xml2js(activity.speak, { compact: false }) as Element;
        } catch(err){
            // Ignore any exceptions. This is effectively a "TryParse", except that XElement doesn't
            // have a TryParse method.
        }

        if (rootElement === undefined || this.getLocalName(rootElement) !== 'speak') {
            // If the text is not valid XML, or if it's not a <speak> node, treat it as plain text.
            rootElement = this.createBaseElement(activity.speak);
        }

        let locale = this.locale;
        if (activity.locale !== undefined && activity.locale.trim().length > 0) {
            try {
                const normalizedLocale: string = activity.locale.toLowerCase();
                if (this.voiceFonts.has(normalizedLocale)){
                    locale = normalizedLocale;
                }
            } catch(err) {
            }
        }

        this.addAttributeIfMissing(rootElement, 'version', '1.0');
        this.addAttributeIfMissing(rootElement, 'xml:lang', `${ locale }`);
        this.addAttributeIfMissing(rootElement, 'xmlns:mstts', 'https://www.w3.org/2001/mstts');

        // Fix issue with 'number_digit' interpreter
        if (rootElement.elements !== undefined && rootElement.elements[0].elements !== undefined) {
            const sayAsElements: Element[] = rootElement.elements[0].elements.filter((e: Element): boolean => e.name === 'say-as');
            sayAsElements.forEach((e: Element): void => {
                this.updateAttributeIfPresent(e, 'interpret-as', 'digits', 'number_digit');
            });
        }

        // add voice element if absent
        const voiceFontOfLocale: string | undefined = this.voiceFonts.get(locale);
        if (voiceFontOfLocale !== undefined && voiceFontOfLocale.trim().length > 0) {
            this.addVoiceElementIfMissing(rootElement, voiceFontOfLocale);
        }

        return js2xml(rootElement, { compact: false });
    }

    /**
     * Add a new attribute to an XML element.
     * @param element The XML element to update.
     * @param attributeName The XML attribute name to add.
     * @param attributeValue The XML attribute value to add.
     */
    private addAttributeIfMissing(element: Element, attributeName: string, attributeValue: string): void {
        if (element.elements !== undefined && element.elements[0] !== undefined) {
            if (element.elements[0].attributes === undefined) {
                element.elements[0].attributes = {
                    attName: attributeValue
                };
            } else {
                if (element.elements[0].attributes[attributeName] === undefined) {
                    element.elements[0].attributes[attributeName] = attributeValue;
                }
            }
        }
    }

    /**
     * Add a new attribute with a voice property to the parent XML element.
     * @param parentElement The XML element to update.
     * @param attributeValue The XML attribute value to add.
     */
    private addVoiceElementIfMissing(parentElement: Element, attributeValue: string): void {
        try {
            if (parentElement.elements === undefined || parentElement.elements[0].elements === undefined) {
                throw new Error('rootElement undefined');
            }
            const existingVoiceElement: Element | undefined = parentElement.elements[0].elements.find((e: Element): boolean => e.name === 'voice');

            // If an existing voice element is undefined (absent), then add it. Otherwise, assume the author has set it correctly.
            if (existingVoiceElement === undefined) {
                const oldElements: Element[] = JSON.parse(JSON.stringify(parentElement.elements[0].elements));
                parentElement.elements[0].elements = [
                    {
                        type: 'element',
                        name: 'voice',
                        attributes: {
                            name: attributeValue
                        },
                        elements: oldElements
                    }
                ];
            } else {
                if (existingVoiceElement.attributes !== undefined) {
                    existingVoiceElement.attributes['name'] = attributeValue;
                }
            }
        } catch (error) {
            throw new Error(`Could not add voice element to speak property: ${ error.message }`);
        }
    }

    /**
     * Update an XML attribute if it already exists.
     * @param element The XML element to update.
     * @param attributeName The XML attribute name to update.
     * @param currentAttributeValue The XML attribute name to update.
     * @param newAttributeValue The XML attribute name to update.
     */
    private updateAttributeIfPresent(element: Element, attributeName: string, currentAttributeValue: string, newAttributeValue: string): void {
        if (element.attributes !== undefined && element.attributes[attributeName] === currentAttributeValue) {
            element.attributes[attributeName] = newAttributeValue;
        }
    }

    private getLocalName(element: Element): string | undefined {
        if (element.elements !== undefined && element.elements.length === 1) {
            return element.elements[0].name;
        }

        return undefined;
    }

    private createBaseElement(value: string): Element {
        // creating simple element
        return {
            elements: [
                {
                    type: 'element',
                    name: 'speak',
                    elements: [
                        {
                            type: 'text',
                            text: value
                        }
                    ],
                    attributes: {
                        xmlns: SetSpeakMiddleware.namespaceURI
                    }
                }
            ]
        };
    }
}

File: sdk/csharp/libraries/microsoft.bot.solutions/Middleware/SetSpeakMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Bot.Solutions.Middleware
{
    using System;
    using System.Collections.Generic;
    using System.Globalization;
    using System.Threading;
    using System.Threading.Tasks;
    using System.Xml;
    using System.Xml.Linq;
    using Microsoft.Bot.Builder;
    using Microsoft.Bot.Schema;
    using Newtonsoft.Json.Linq;

    /// <summary>
    /// Set Speech Synthesis Markup Language (SSML) on an Activity's Speak property with locale and voice input.
    /// </summary>
    public class SetSpeakMiddleware : IMiddleware
    {
        private const string DefaultLocale = "en-US";

        private static readonly IDictionary<string, string> DefaultVoiceFonts = new Dictionary<string, string>()
        {
            { "de-DE", "Microsoft Server Speech Text to Speech Voice (de-DE, Hedda)" },
            { "en-US", "Microsoft Server Speech Text to Speech Voice (en-US, Jessa24kRUS)" },
            { "es-ES", "Microsoft Server Speech Text to Speech Voice (es-ES, Laura, Apollo)" },
            { "fr-FR", "Microsoft Server Speech Text to Speech Voice (fr-FR, Julie, Apollo)" },
            { "it-IT", "Microsoft Server Speech Text to Speech Voice (it-IT, Cosimo, Apollo)" },
            { "zh-CN", "Microsoft Server Speech Text to Speech Voice (zh-CN, HuihuiRUS)" },
        };

        private static readonly ISet<string> DefaultChannels = new HashSet<string>()
        {
            Connector.Channels.DirectlineSpeech,
            Connector.Channels.Emulator,
        };

        private static string _locale;

        private static IDictionary<string, string> _voiceFonts;

        private static ISet<string> _channels;

        private static readonly XNamespace NamespaceURI = @"https://www.w3.org/2001/10/synthesis";

        /// <summary>
        /// Initializes a new instance of the <see cref="SetSpeakMiddleware"/> class.
        /// </summary>
        /// <param name="locale">If null, use en-US.</param>
        /// <param name="voiceFonts">Map voice font for locale like en-US to "Microsoft Server Speech Text to Speech Voice (en-US, Jessa24kRUS)".</param>
        /// <param name="channels">Set SSML for these channels. If null, use <see cref="Connector.Channels.DirectlineSpeech"/> and <see cref="Connector.Channels.Emulator"/>.</param>
        public SetSpeakMiddleware(string locale = null, IDictionary<string, string> voiceFonts = null, ISet<string> channels = null)
        {
            _locale = locale ?? DefaultLocale;
            _voiceFonts = voiceFonts ?? DefaultVoiceFonts;
            _channels = channels ?? DefaultChannels;
        }

        /// <summary>
        /// If outgoing Activities are messages and using one of the desired channels, decorate the Speak property with an SSML formatted string.
        /// </summary>
        /// <param name="context">The Bot Context object.</param>
        /// <param name="next">The next middleware component to run.</param>
        /// <param name="cancellationToken">The cancellation token for the task.</param>
        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
        public Task OnTurnAsync(ITurnContext context, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            context.OnSendActivities(async (ctx, activities, nextSend) =>
            {
                foreach (var activity in activities)
                {
                    switch (activity.Type)
                    {
                        case ActivityTypes.Message:
                            activity.Speak = GetActivitySpeakText(activity);

                            if (_channels.Contains(activity.ChannelId))
                            {
                                activity.Speak = DecorateSSML(activity);
                            }

                            break;
                    }
                }

                return await nextSend().ConfigureAwait(false);
            });

            return next(cancellationToken);
        }

        /// <summary>
        /// Gets the speak text for the activity.
        /// </summary>
        /// <param name="activity">Outgoing bot Activity.</param>
        /// <returns>speech text string value.</returns>
        private static string GetActivitySpeakText(Activity activity)
        {
            // return speak or text value if they already exist in the activity
            var result = activity.Speak ?? activity.Text;
            if (result != null)
            {
                return result;
            }

            // return speak value of first attachment if an attachment exists and has a speak value
            if (activity.Attachments.Count > 0)
            {
                var attachmentContent = activity.Attachments[0].Content;
                if (attachmentContent != null)
                {
                    var contentObj = attachmentContent as JObject;
                    return contentObj?["speak"]?.ToString();
                }
            }

            return null;
        }

            /// <summary>
            /// Formats an existing string to be formatted for Speech Synthesis Markup Language with a voice font.
            /// </summary>
            /// <param name="activity">Outgoing bot Activity.</param>
            /// <returns>SSML-formatted string to be used with synthetic speech.</returns>
            private static string DecorateSSML(Activity activity)
        {
            if (string.IsNullOrWhiteSpace(activity.Speak))
            {
                return activity.Speak?.Trim();
            }

            XElement rootElement = null;
            try
            {
                rootElement = XElement.Parse(activity.Speak);
            }
            catch (XmlException)
            {
                // Ignore any exceptions. This is effectively a "TryParse", except that XElement doesn't
                // have a TryParse method.
            }

            if (rootElement == null || rootElement.Name.LocalName != "speak")
            {
                // If the text is not valid XML, or if it's not a <speak> node, treat it as plain text.
                rootElement = new XElement(NamespaceURI + "speak", activity.Speak);
            }

            var locale = _locale;
            if (!string.IsNullOrEmpty(activity.Locale))
            {
                try
                {
                    var normalizedLocale = new CultureInfo(activity.Locale).Name;
                    if (_voiceFonts.ContainsKey(normalizedLocale))
                    {
                        locale = normalizedLocale;
                    }
                }
                catch (CultureNotFoundException)
                {
                }
            }

            AddAttributeIfMissing(rootElement, "version", "1.0");
            AddAttributeIfMissing(rootElement, XNamespace.Xml + "lang", locale);
            AddAttributeIfMissing(rootElement, XNamespace.Xmlns + "mstts", "https://www.w3.org/2001/mstts");

            var sayAsElements = rootElement.Elements("say-as");
            foreach (var element in sayAsElements)
            {
                UpdateAttributeIfPresent(element, "interpret-as", "digits", "number_digit");
            }

            // add voice element if absent
            AddVoiceElementIfMissing(rootElement, _voiceFonts[locale]);

            return rootElement.ToString(SaveOptions.DisableFormatting);
        }

        /// <summary>
        /// Add a new attribute to an XML element.
        /// </summary>
        /// <param name="element">The XML element to update.</param>
        /// <param name="attributeName">The XML attribute name to add.</param>
        /// <param name="attributeValue">The XML attribute value to add.</param>
        private static void AddAttributeIfMissing(XElement element, XName attributeName, string attributeValue)
        {
            var existingAttribute = element.Attribute(attributeName);
            if (existingAttribute == null)
            {
                element.Add(new XAttribute(attributeName, attributeValue));
            }
        }

        /// <summary>
        /// Add a new attribute with a voice property to the parent XML element.
        /// </summary>
        /// <param name="parentElement">The XML element to update.</param>
        /// <param name="attributeValue">The XML attribute to add.</param>
        private static void AddVoiceElementIfMissing(XElement parentElement, string attributeValue)
        {
            try
            {
                var existingVoiceElement = parentElement.Element("voice") ?? parentElement.Element(NamespaceURI + "voice");

                // If an existing voice element is null (absent), then add it. Otherwise, assume the author has set it correctly.
                if (existingVoiceElement == null)
                {
                    var existingNodes = parentElement.DescendantNodes();

                    XElement voiceElement = new XElement("voice", new XAttribute("name", attributeValue));
                    voiceElement.Add(existingNodes);
                    parentElement.RemoveNodes();
                    parentElement.Add(voiceElement);
                }
                else
                {
                    existingVoiceElement.SetAttributeValue("name", attributeValue);
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Could not add voice element to speak property {ex.Message}");
            }
        }

        /// <summary>
        /// Update an XML attribute if it already exists.
        /// </summary>
        /// <param name="element">The XML element to update.</param>
        /// <param name="attributeName">The XML attribute name to update.</param>
        /// <param name="currentAttributeValue">The current XML attribute's value.</param>
        /// <param name="newAttributeValue">The new XML attribute's value.</param>
        private static void UpdateAttributeIfPresent(XElement element, XName attributeName, string currentAttributeValue, string newAttributeValue)
        {
            var existingAttribute = element?.Attribute(attributeName);
            if (existingAttribute?.Value == currentAttributeValue)
            {
                existingAttribute?.SetValue(newAttributeValue);
            }
        }
    }
}